jaeger-core/src/main/java/com/uber/jaeger/samplers/PerOperationSampler.java

```diff
@@ -0,0 +1,120 @@
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import lombok.AccessLevel;
+import lombok.Data;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Computes {@link #sample(String, long)} using the name of the operation, and maintains a specific
+ * {@link GuaranteedThroughputSampler} instance for each operation.
+ */
+@Slf4j
+@Data
+@Getter(AccessLevel.NONE)
+public class PerOperationSampler implements Sampler {
+  private final ConcurrentHashMap<String, GuaranteedThroughputSampler> operationNameToSampler;
+  private final int maxOperations;
+  private volatile ProbabilisticSampler defaultSampler;
+
+  public PerOperationSampler(int maxOperations, OperationSamplingParameters strategies) {
+    this(maxOperations,
+         new ConcurrentHashMap<String, GuaranteedThroughputSampler>(),
+         new ProbabilisticSampler(strategies.getDefaultSamplingProbability()));
+    update(strategies);
+  }
+
+  // Visible for testing
+  PerOperationSampler(int maxOperations,
+                      ConcurrentHashMap<String, GuaranteedThroughputSampler> operationNameToSampler,
+                      ProbabilisticSampler defaultSampler) {
+
+    this.operationNameToSampler = operationNameToSampler;
+    this.maxOperations = maxOperations;
+    this.defaultSampler = defaultSampler;
+  }
+
+  /**
+   * Updates the probabilistic samplers for each operation
+   * @param strategies The parameters for operation sampling
+   * @return true iff any samplers were updated
+   */
+  public synchronized boolean update(OperationSamplingParameters strategies) {
+    boolean isUpdated = false;
+
+    double lowerBound = strategies.getDefaultLowerBoundTracesPerSecond();
+    ProbabilisticSampler defaultSampler = new ProbabilisticSampler(strategies.getDefaultSamplingProbability());
+
+    if (!defaultSampler.equals(this.defaultSampler)) {
+      this.defaultSampler = defaultSampler;
+      isUpdated = true;
+    }
+
+    for (PerOperationSamplingParameters strategy : strategies.getPerOperationStrategies()) {
+      String operation = strategy.getOperation();
+      double samplingRate = strategy.getProbabilisticSampling().getSamplingRate();
+      GuaranteedThroughputSampler sampler = new GuaranteedThroughputSampler(samplingRate, lowerBound);
+      GuaranteedThroughputSampler oldSampler = operationNameToSampler.get(operation);
+
+      if (oldSampler == null) {
+        if (operationNameToSampler.size() < maxOperations) {
+          operationNameToSampler.put(operation, sampler);
+          isUpdated = true;
+        } else {
+          log.info("Exceeded the maximum number of operations({}) for per operations sampling",
+                   maxOperations);
+        }
+      } else if (!sampler.equals(oldSampler)) {
+        operationNameToSampler.replace(operation, sampler);
+        isUpdated = true;
+      }
+    }
+
+    return isUpdated;
+  }
+
+  @Override
+  public SamplingStatus sample(String operation, long id) {
```

Comment:
> As such, we are setting a default of a GTSampler for endpoints that we haven't seen yet when we have space in the map, but we use a default of a Probabilistic sampler otherwise. I don't see the benefit of doing this.

This is one of the objective of the project, to use GTSampler for each endpoint to guarantee minimum rate. Default probabilistic sampler (a) is not going to provide that guarantee, and (b) is only used when service exceeds the limit on the number of operations we want to track, which should be never. GTSampler as default is not any better when applied to random endpoints, and is largely irrelevant because of (b) above.